fix: reset cache for timezone changes

troZee opened the pull request:
## Summary
A new native function resetTimezoneCache has been added to HermesRuntime object. This function handles resetting the timezone cache, while the timezone is changed

This fix has been implemented based on the @tmikov recommendation:
 
```
We propose adding a JSI method to HermesRuntime for reseting the TZ cache. The integrator (React Native) can then call this method after detecting that the TZ has changed, using whatever approach is suitable for the platform.
```

## Test Plan 

1. Add the below code to `include/hermes/VM/JSLib/DateCache.h`
```
  void reset() {
    std::cout << "Calling reset" << std::endl;
```    

2. Build Hermes 
3. Run the unit tests and check if this method is called 

## How to reproduce it?

1. `bunx @react-native-community/cli@latest init ReactNativePlayground769 --version 0.76.9`
2.  Replace the App.tsx with the below code 
```
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, {useEffect, useState} from 'react';
import type {PropsWithChildren} from 'react';
import {
  AppState,
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
} from 'react-native';

import {
  Colors,
  DebugInstructions,
  Header,
  LearnMoreLinks,
  ReloadInstructions,
} from 'react-native/Libraries/NewAppScreen';

import * as RNLocalize from 'react-native-localize';

const useTimezoneChange = () => {
  const [timezone, setTimezone] = useState(RNLocalize.getTimeZone());

  useEffect(() => {
    // Set up listener for timezone changes
    const subscription = AppState.addEventListener('change', nextAppState => {
      if (nextAppState === 'active') {
        const newTimezone = RNLocalize.getTimeZone();
        if (newTimezone !== timezone) {
          setTimezone(newTimezone);
          // Handle timezone change here
          console.log('Timezone changed to:', newTimezone);
        }
      }
    });

    // Clean up listener on unmount
    return () => {
      subscription.remove();
    };
  }, [timezone]);

  return timezone;
};

const TimezoneAwareComponent = () => {
  const currentTimezone = useTimezoneChange();

  const date = new Date(
    'Mon Apr 28 2025 22:05:36 GMT+0900 (Japan Standard Time)',
  );
  date.getTimezoneOffset();
  return (
    <View>
      <Text>Current timezone: {currentTimezone}</Text>
      <Text>
        Date: {date.getHours()}:{date.getMinutes()}
      </Text>
    </View>
  );
};

const DateComponent = () => {
  const date = new Date(
    'Mon Apr 28 2025 22:05:36 GMT+0900 (Japan Standard Time)',
  );
  return (
    <View>
      <Text>
        Fixed Date: {date.getHours()}:{date.getMinutes()}
      </Text>
    </View>
  );
};

function App(): React.JSX.Element {
  const isDarkMode = useColorScheme() === 'dark';

  const backgroundStyle = {
    backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
  };

  return (
    <SafeAreaView style={backgroundStyle}>
      <StatusBar
        barStyle={isDarkMode ? 'light-content' : 'dark-content'}
        backgroundColor={backgroundStyle.backgroundColor}
      />
      <ScrollView
        contentInsetAdjustmentBehavior="automatic"
        style={backgroundStyle}>
        <Header />
        <View
          style={{
            backgroundColor: isDarkMode ? Colors.black : Colors.white,
          }}>
          <TimezoneAwareComponent />
          <DateComponent />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
  },
  highlight: {
    fontWeight: '700',
  },
});

export default App;
```

## Video


https://github.com/user-attachments/assets/68266155-02c2-4d3d-bd6f-60a07a0d3b6d


https://github.com/user-attachments/assets/290d114d-ed55-49dd-80e1-fc98d18bfae6

tmikov commented:
Thank you for submitting this PR. However I don't really understand this change. It seems it is exposing a new non-standard method `resetCache()`. Why is that desirable? It doesn't seem this method is ever called, so how does it address the problem?

lavenzg commented:
I can't reproduce your result locally. This is the result I'm getting:
```
TZ=America/Santiago ~/hermes_build/build/bin/hermes gh1693.js
Fri Sep 01 2023 00:00:00 GMT-0400
Sat Sep 02 2023 00:00:00 GMT-0400
Sun Sep 03 2023 01:00:00 GMT-0300
Mon Sep 04 2023 00:00:00 GMT-0300
```

We only need to reset cache if the timezone changes when hermes is running (e.g., set the TZ environment variable). Is the "hermes playground" an App that is already running before you change the timezone in step 1?

troZee commented:
@tmikov 
```
It seems it is exposing a new non-standard method resetCache() to JS
```

Yes correct. I forgot to mention that in the PR description, but when we call that function for every calendar event, which comes from the API all, it works and those timezone changes are reflected on the UI. I will update the PR description accordingly 

`Why is that desirable?`
In our app, we might have a situation where the timezone is changed (eg. someone is traveling from Tokyo to Sydney). When the timezone is changed, it should display a correct offset and calculate that based on the changed timezone 

`It doesn't seem this method is ever called, so how does it address the problem?`

I should have mentioned that in the PR description, that we call that method on the React Native side. 

`Also, what is the "hermes playground"?`
I meant hermes cli from https://github.com/facebook/hermes/releases/tag/v0.13.0 

@lavenzg 
I answered on all question above, but I will put more comment to your questions 

```
 Is the "hermes playground" an App that is already running before you change the timezone in step 1?
 ```
 Those changes were tested inside the RN app and we called that exposed function on the react-native side

mdjastrzebski commented:
Together with @troZee  we are looking for a way to make Date work correctly in case underlying TZ changes. It seems that the caching has been introduced around 0.76 in this [PR](https://github.com/facebook/hermes/commit/59ea64c52d78ecd8e281e576fab16c208f87d4ae).

I see two ways that could be done:
1. Hermes will detect TZ change and invalidate the cache itself
2. Hermes will expose method to invalidate the cache manually.

@tmikov wdyt?

tmikov commented:
So, the idea is one of:
a) to introduce a non-standard JS API to reset the TZ cache
b) introduce a native API for it
c) have Hermes correctly detect that the TZ has changed.

If we decide to go with a), it would likely be a method on HermesInternal, not on `Date`.

In your experiments, when are you calling `resetCache()`?

tmikov commented:
We discussed it internally and concluded that the best way to handle this is to detect a timezone change using platform-specific APIs and then reset the cache. At the same time, we don't think hooking into platform APIs should be done by Hermes - it may even be impossible.

We propose adding a JSI method to HermesRuntime for reseting the TZ cache. The integrator (React Native) can then call this method after detecting that the TZ has changed, using whatever approach is suitable for the platform.

troZee commented:
@tmikov @lavenzg 
Thank you for checking that issue 🙏 

```
Do you want to submit a PR adding it? It just need to call localTimeOffsetCache.reset().
```

Should I submit a new PR or apply changes to this one?

tmikov commented:
> Should I submit a new PR or apply changes to this one?

Makes no difference either way.

troZee commented:
@tmikov @lavenzg 
I have updated the code. Please let me know if I should test additional scenarios. 

I did not add any unit tests, since I am not sure if those are needed for the exposed methods to JS

troZee commented:
@tmikov @lavenzg 

We noticed the following comment https://github.com/facebook/hermes/pull/1693/files#diff-cd87104dd09a2271bbedfe3f5aab799f6efa6c1583c8dab9f1abdc1a57297b4cR202 I changed the implementation based on this comment, so instead of exposing the `reset` method, I created a new method `needsToReset()` for changing `needsToReset_` flag to make this function thread-safe

tmikov commented:
Hi, I am sorry, as I said earlier, this is what we are proposing:

> We propose adding a JSI method to HermesRuntime for reseting the TZ cache. The integrator (React Native) can then call this method after detecting that the TZ has changed, using whatever approach is suitable for the platform.

troZee commented:
@tmikov reverted

troZee commented:
@tmikov @lavenzg 
I added more unit tests for that PR. I run the below command to check if all unit tests are passing 

```
hermes git:(fix/1121-tz-change) ✗ cmake --build ../build --target check-hermes
[0/1] Running the Hermes regression tests
Testing Time: 38.50s
  Expected Passes    : 2238
  Unsupported Tests: 44
```

Could you also run the CI for this PR? I would like to be sure everything has been addressed in this PR
<img width="497" alt="Screenshot 2025-05-08 at 22 54 43" src="https://github.com/user-attachments/assets/0eba6c5f-55ee-4c5b-8806-565b2f7d0f15" />

Please let me know if something is still missing in this PR

troZee commented:
@tmikov @lavenzg 
I moved this implementation from HermesInternal to HermesRuntime. Please let me know if it is implemented correctly. 

I need help with the unit tests here https://github.com/facebook/hermes/pull/1693/files#diff-dac96edcb6b8bf138649480205106b295f2fbc8768bdd85541b8624d070527d6R107 . For some reason, there is a linker error, and those files are in the CMAKE.txt file. I am unsure why those files are not linked correctly when running unit tests. 

If this comment part is not needed, then I can remove that

troZee commented:
@tsaichien 
In this commit https://github.com/facebook/hermes/commit/2bba5f049b6b9c3e959f92391cb0f682c70d668b
the method getVMRuntimeUnsafe changed the returned value from`::hermes::vm::Runtime` to `void`:

```
- ::hermes::vm::Runtime *getVMRuntimeUnsafe() const override;
+ void *getVMRuntimeUnsafe() const override;
 ```
 
 Could you advise me on what I should do to resolve the conflicts for this PR?

troZee commented:
@tmikov @lavenzg 
The conflicts are resolved. The PR is ready for review

troZee commented:
@tmikov @lavenzg
I've addressed all the points raised in the initial review and believe it's ready for final approval.

Could you please provide an estimated timeline for when I can expect the review to be completed? This will help with planning.

Thanks!

tmikov commented on API/hermes/hermes.h:
```diff
@@ -179,6 +179,9 @@ class HERMES_EXPORT IHermes : public jsi::ICast {
   /// be used by third parties.
   virtual sampling_profiler::Profile dumpSampledTraceToProfile() = 0;
 
+  /// Reset timezone cache
```
Can you please add a much more detailed comment explaining what purpose this call serves, uses cases when it should be invoked, etc.

tmikov commented:
Landing changes to JSI is a complex and time consuming process, unfortunately. It will take us some time.

tmikov commented:
Merged in https://github.com/facebook/hermes/commit/1b4de35e4e42c807393b35f26892a4e531255b39
